torch/nn/modules/module.py

```diff
@@ -193,13 +193,40 @@ def _apply(self, fn):
         for module in self.children():
             module._apply(fn)
 
-        for param in self._parameters.values():
+        def compute_should_use_data(tensor, tensor_applied):
+            if torch._has_same_tensorimpl_type(tensor, tensor_applied):
+                # If the new tensor has the same TensorImpl type as the existing tensor,
+                # then we take `torch.__future__.get_overwrite_module_params_on_conversion()`
+                # into account.
+                return not torch.__future__.get_overwrite_module_params_on_conversion()
+            else:
+                # If the new tensor has a different TensorImpl type, we always overwrite
+                # the existing tensor.
+                return False
+
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                should_use_data = compute_should_use_data(param, fn(torch.empty(1, dtype=param.dtype, layout=param.layout, device=param.device)))
+                if should_use_data:
+                    # Tensors stored in modules are graph leaves, and we don't
+                    # want to create copy nodes, so we have to unpack the data.
+                    param.data = fn(param.data)
+                else:
+                    # Tensors stored in modules are graph leaves, and we don't want to
+                    # track autograd history of `param_applied`, so we have to use
+                    # `with torch.no_grad():`
+                    with torch.no_grad():
+                        param_applied = fn(param)
+                    self._parameters[key] = Parameter(param_applied, param.requires_grad)
+                    assert isinstance(self._parameters[key], Parameter)
```

Comment:
can you switch the order of the assert and the assignment?  You usually check if something is valid before doing it.

Reply:
Fixed.